File: skellig-connection-cassandra/src/main/java/org/skellig/connection/cassandra/CassandraSelectRequestExecutor.java
package org.skellig.connection.cassandra;

import com.datastax.driver.core.PreparedStatement;
import com.datastax.driver.core.ResultSet;
import com.datastax.driver.core.Session;
import org.skellig.connection.database.BaseDatabaseRequestExecutor;
import org.skellig.connection.database.exception.DatabaseChannelException;
import org.skellig.connection.database.model.DatabaseRequest;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

class CassandraSelectRequestExecutor extends BaseDatabaseRequestExecutor {

    private static final String COMPARATOR = "comparator";

    private Session session;

    public CassandraSelectRequestExecutor(Session session) {
        this.session = session;
    }

    @Override
    public Object execute(DatabaseRequest databaseRequest) {
        try {
            if (databaseRequest.getQuery() != null) {
                return session.execute(databaseRequest.getQuery());
            } else {
                Map<String, Object> searchCriteria = databaseRequest.getColumnValuePairs().orElse(Collections.emptyMap());
                String query = composeFindQuery(databaseRequest, searchCriteria);

                PreparedStatement preparedStatement = session.prepare(query);
                Object[] rawParameters = convertToRawParameters(searchCriteria);

                return extractFromResultSet(session.execute(preparedStatement.bind(rawParameters)));
            }

        } catch (Exception ex) {
            throw new DatabaseChannelException(ex.getMessage(), ex);
        }
    }

    private List<Map<String, Object>> extractFromResultSet(ResultSet resultSet) {
        List<Map<String, Object>> result = new ArrayList<>();

        resultSet.forEach(row -> {
            Map<String, Object> resultRow = new LinkedHashMap<>();
            result.add(resultRow);
            row.getColumnDefinitions().asList()
                    .forEach(column -> resultRow.put(column.getName(), row.getObject(column.getName())));
        });
        return result;
    }

    private Object[] convertToRawParameters(Map<String, Object> searchCriteria) {
        return searchCriteria.values().stream()
                .map(this::getParameterValue)
                .toArray();
    }

    private String composeFindQuery(DatabaseRequest databaseRequest, Map<String, Object> searchCriteria) {
        StringBuilder queryBuilder = new StringBuilder();
        queryBuilder.append("SELECT * FROM ");
        queryBuilder.append(databaseRequest.getTable());
        if (!searchCriteria.isEmpty()) {
            queryBuilder.append(" WHERE ");
            String columns = searchCriteria.entrySet().stream()
                    .map(entry -> {
                        if (entry.getValue() instanceof Map) {
                            String comparator = String.valueOf(((Map) entry.getValue()).get(COMPARATOR));
                            return entry.getKey() + comparator + " ?";
                        } else {
                            return entry.getKey() + getCompareOperator(entry.getValue()) + "?";
                        }
                    })
                    .collect(Collectors.joining(" AND "));

            queryBuilder.append(columns);
        }
        return queryBuilder.toString();
    }

    private String getCompareOperator(Object valueToCompare) {
        return String.valueOf(valueToCompare).contains("%") ? " like " : " = ";
    }
}

File: skellig-test-step-processing-cassandra/src/main/kotlin/org/skellig/teststep/processor/cassandra/CassandraSelectRequestExecutor.kt
package org.skellig.teststep.processor.cassandra

import com.datastax.oss.driver.api.core.CqlSession
import com.datastax.oss.driver.api.core.cql.ResultSet
import com.datastax.oss.driver.api.core.cql.Row
import com.datastax.oss.driver.api.core.cql.SimpleStatement
import org.skellig.teststep.processing.exception.TestStepProcessingException
import org.skellig.teststep.processing.util.debug
import org.skellig.teststep.processing.util.logger
import org.skellig.teststep.processor.db.model.DatabaseRequest
import java.util.function.Consumer

/**
 * This class is responsible for executing Cassandra select queries on the database using.
 *
 * @property session The Cassandra session used to execute queries in the database.
 */
internal open class CassandraSelectRequestExecutor(private val session: CqlSession) : BaseCassandraRequestExecutor() {

    companion object {
        private const val COMPARATOR = "comparator"
    }

    private val log = logger<CassandraSelectRequestExecutor>()

    override fun execute(databaseRequest: DatabaseRequest): Any {
        return try {
            if (databaseRequest.query != null) {
                executeQuery(databaseRequest.query!!,
                    databaseRequest.queryParameters
                        ?.map { getParameterValue(it) }
                        ?.toTypedArray() ?: emptyArray())
            } else {
                val searchCriteria = databaseRequest.columnValuePairs ?: emptyMap()
                val query = composeFindQuery(databaseRequest, searchCriteria)
                executeQuery(query, convertToRawParameters(searchCriteria))
            }
        } catch (ex: Exception) {
            throw TestStepProcessingException(ex.message, ex)
        }
    }

    private fun executeQuery(query: String, queryParameters: Array<Any?>): Any {
        val response = extractFromResultSet(session.execute(SimpleStatement.newInstance(query, *queryParameters)))
        log.debug {
            "Query has been executed successfully: $query " +
                    "with parameters: ${queryParameters.contentToString()} " +
                    "and response: $response"
        }
        return response
    }

    private fun extractFromResultSet(resultSet: ResultSet): List<Map<String, Any?>> {
        val result: MutableList<Map<String, Any?>> = ArrayList()
        resultSet.forEach(Consumer { row: Row ->
            val resultRow: MutableMap<String, Any?> = LinkedHashMap()
            result.add(resultRow)
            row.columnDefinitions
                .forEach(Consumer { column ->
                    resultRow[column.name.asInternal()] = row.getObject(column.name)
                })
        })
        return result
    }

    private fun convertToRawParameters(searchCriteria: Map<String, Any?>): Array<Any?> {
        return searchCriteria.values.stream()
            .map { getParameterValue(it) }
            .toArray()
    }

    private fun composeFindQuery(databaseRequest: DatabaseRequest, searchCriteria: Map<String, Any?>): String {
        val queryBuilder = StringBuilder()
        queryBuilder.append("SELECT * FROM ")
        queryBuilder.append(databaseRequest.table)
        if (searchCriteria.isNotEmpty()) {
            queryBuilder.append(" WHERE ")
            val columns = searchCriteria.entries
                .joinToString(" AND ") {
                    if (it.value is Map<*, *>) {
                        val comparator = (it.value as Map<*, *>)[COMPARATOR].toString()
                        it.key + comparator + " ?"
                    } else {
                        it.key + getCompareOperator(it.value) + "?"
                    }
                }
            queryBuilder.append(columns)
        }
        return queryBuilder.append(" ALLOW FILTERING").toString()
    }

    private fun getCompareOperator(valueToCompare: Any?): String {
        return if (valueToCompare.toString().contains("%")) " like " else " = "
    }
}